### PR TITLE
Fix runtime manager's Makefiles across workspaces

### PR DESCRIPTION
**Pull request by gbryant-arm**
Currently the build system doesn't recompile the runtime manager when its dependencies, e.g. `transport-protocol`, are modified.
This makes debugging tedious and error prone: many times I was trying to figure out the reason for network issues or crashes, before realising the binaries had not been recompiled.  @egrimley-arm faced similar issues.
This PR fixes that by calling Cargo to look for changes in the source of  runtime-manager and all its dependencies, instead of relying on traditional Makefile targets.

### CODE REVIEW

File: runtime-manager/Makefile
Comment (egrimley-arm) [inline runtime-manager/Makefile:44]: You should add `build-nitro` to the `.PHONY` dependencies.

**Comment by dreemkiller** [comment]
Cargo, even when it's doing nothing, it a relatively slow process (I think 0.5s to check dependencies).
Make is much faster. Is it possible to modify the Make solution to pick up the relevant changes instead? If not, this will work.

**Comment by gbryant-arm** [comment]
Agreed, Cargo is slow, but mainly at compiling, which we can't avoid, right?
Plus it automatically keeps track of the sources, fetches updates for the crate's dependencies, etc.
I don't think using Cargo here will add much overhead to the build process

**Comment by gbryant-arm** [comment]
AFAIK the RM is the only component of the build (or at least the only one I had issues with) that is currently relying on Make to detect changes, so that would be 0.5 additional seconds per build, which seems acceptable.
The main Makefile directly calls cargo for every other component, and I don't think we want to change that, right?

**Review by mathias-arm — CHANGES_REQUESTED** [review]
This `Makefile` is destined to be removed. Look at the ones in `workspaces/linux-runtime` and `workspaces/nitro-runtime`

File: workspaces/linux-runtime/Makefile
Comment (mathias-arm) [inline workspaces/linux-runtime/Makefile:31]: This does not seem to be right. 
`runtime-manager-enclave` rule builds `target/$(PROFILE_PATH)/runtime_manager_enclave `, and it's the latter that should be copied to `css-linux.bin`

**Comment by mathias-arm** [comment]
@gbryant-arm: regarding @dreemkiller's comments, have you checked that with multiple `make` invocations do not result in superfluous rebuilds (i.e. `cargo` figures out there is no need to recompile).

**Comment by dominic-mulligan-arm** [comment]
What is the status of this PR @gbryant-arm?

**Comment by gbryant-arm** [comment]
WIP

**Comment by dominic-mulligan-arm** [comment]
Converted the PR to a draft, in that case.

**Comment by mathias-arm** [comment]
Looking at the nitro changes, it seems `nitro-cli` might be run even without any changes to the `runtime-manager` which brings me closer to @dreemkiller's concerns. Looking back at the original issue, what kind of changes were not picked up?

**Comment by mathias-arm** [comment]
Looking at the nitro changes, it seems `nitro-cli` might be run even without any changes to the `runtime-manager` which brings me closer to @dreemkiller's concerns. Looking back at the original issue, what kind of changes were not picked up?

**Comment by gbryant-arm** [comment]
> Looking back at the original issue, what kind of changes were not picked up?

For example, `runtime manager` doesn't get recompiled when `transport protocol` is changed.


> nitro-cli might be run even without any changes to the runtime-manager

I can't find `nitro-cli` anywhere. Does that target still exist in the workspace Makefiles?

**Comment by gbryant-arm** [comment]
> @gbryant-arm: regarding @dreemkiller's comments, have you checked that with multiple make invocations do not result in superfluous rebuilds (i.e. cargo figures out there is no need to recompile).

Checked. Takes cargo about 0.2s to look for changes and realize there's nothing to build.

**Comment by mathias-arm** [comment]
For the nitro part, it makes sense to re-run `cargo build` each time. I am not sure what now happens with the rules for `runtime_manager.eif` and `PCR0`, does `nitro-cli` get called appropriately?

**Comment by gbryant-arm** [comment]
> `runtime_manager.eif` and `PCR0`, does `nitro-cli`

`nitro-cli` gets called whenever the `runtime_manager.eif` or (and?) `PCR0` targets get executed. The CI logs confirm that

**Comment by mathias-arm** [comment]
> > `runtime_manager.eif` and `PCR0`, does `nitro-cli`
> 
> `nitro-cli` gets called whenever the `runtime_manager.eif` or (and?) `PCR0` targets get executed. The CI logs confirm that

The issue I was thinking about is on subsequent builds, if those rules are called regardless of whether `runtime_manager_enclave` was updated.

**Comment by mathias-arm** [comment]
> One hacky way to tell is to check the creation date of the produced binaries. Ugly

This is one of the main feature of Make.

**Comment by gbryant-arm** [comment]
> This is one of the main features of Make.

It is, but I don't think we can use that here, since the target is already used. We'd have a cycle of targets

**Comment by gbryant-arm** [comment]
The goal of this PR is to replace Make with cargo to track Rust-related dependencies

**Comment by gbryant-arm** [comment]
@mathias-arm Are you happy with this PR? I addressed your request for changes.

**Comment by dominic-mulligan-arm** [comment]
What is the status of this PR?

**Comment by gbryant-arm** [comment]
> What is the status of this PR?
Waiting for @mathias-arm 's approval. It needs rebasing too

**Comment by dominic-mulligan-arm** [comment]
Can you go ahead and rebase in the meantime?  @mathias-arm are you happy with this PR?  @gbryant-arm claims to have addressed all concerns.

**Comment by mathias-arm** [comment]
#339 and #375 are co-dependent because the second chunk of [b1939e4](https://github.com/veracruz-project/veracruz/pull/375/commits/b1939e4c19200e7b9251e13d3ff06730ec90ac52) relies on [0ac34f3](https://github.com/veracruz-project/veracruz/pull/339/commits/0ac34f3368039d3b8a1788feb81b6db205d8c86c). I think both should be removed from their respective PR.

**Comment by dominic-mulligan-arm** [comment]
OK, trying to break the deadlock because this has been dragging on for way too long:

- @mathias-arm: can you be more specific about what's wrong with 14d5006?
- @gbryant-arm: please combine PRs #339 and #375 into a single PR.  Also, can you be more specific about why this change is needed, and why the same effect cannot be obtained using `make`?  That's not clear at all from the thread above, just a comment about "circular targets".  Isn't one of the main features of `make` to detect changes in files and act on them?

**Comment by mathias-arm** [comment]
Regarding https://github.com/veracruz-project/veracruz/commit/14d5006060d6763042dcb1acc1de62c0157e0878 it's a commit titled `Fix Makefile` that deletes the `docker` directory. I assume this is a mistake which happened because `docker` switched from being a submodule to a regular directory.

My take on @gbryant-arm's change is that in order to be sure things are properly rebuilt, we switch from expressing dependencies on the existence of files to phony targets that will force cargo invocations (which should be quick hopefully). Even if @dreemkiller and I have expressed concern about the consequence on build times, the principle is sound. My remaining objections is that the two PRs rely on `$(WORKSPACE_DIR)/linux-runtime/target/$(PROFILE_PATH)/runtime_manager_enclave` being copied to `$(WORKSPACE_DIR)/linux-runtime/runtime_manager_enclave` which I think is superfluous and confusing.

**Comment by gbryant-arm** [comment]
I'm fixing https://github.com/veracruz-project/veracruz/issues/383 for Linux and Nitro as well as this is related

**Comment by gbryant-arm** [comment]
Okay ready for review

**Review by mathias-arm — APPROVED** [review]

**Pull request merged**
